src/dlgIRC.cpp

```diff
@@ -20,306 +22,715 @@
+    connection->setHost(mHostName);
+    connection->setPort(mHostPort);
+}
 
-    if (nick.isEmpty()) {
-        nick = tr("Mudlet%1").arg(QString::number(rand() % 10000));
-        QFile file(QDir::homePath() + "/.config/mudlet/irc_nick");
-        file.open(QIODevice::WriteOnly | QIODevice::Unbuffered);
-        QDataStream ofs(&file);
-        ofs << nick;
-        file.close();
+dlgIRC::~dlgIRC()
+{
+    if (connection->isActive()) {
+        const QString quitMsg = tr("%1 closed their client.").arg(mNickName);
+        connection->quit(quitMsg);
+        connection->close();
+    }
+}
+
+void dlgIRC::startClient()
+{
+    if (mIrcStarted) {
+        return;
     }
 
-    connection->setNickName(nick);
-    mNick = nick;
-    connection->setUserName("mudlet");
-    connection->setRealName(mudlet::self()->version);
-    connection->setHost("irc.freenode.net");
-    connection->setPort(6667);
+    connection->sendCommand(IrcCommand::createJoin(mChannels));
     connection->open();
+
+    setupBuffers();
+
+    ircBrowser->append(IrcMessageFormatter::formatMessage(tr("$ Statring Mudlet IRC Client...")));
+    ircBrowser->append(IrcMessageFormatter::formatMessage(tr("$ Host: %1:%2").arg(mHostName, QString::number(mHostPort))));
+    ircBrowser->append(IrcMessageFormatter::formatMessage(tr("$ Nick: %1").arg(mNickName)));
+    ircBrowser->append(IrcMessageFormatter::formatMessage(tr("$ Auto-Join Channels: %1").arg(mChannels.join(" "))));
+    ircBrowser->append(IrcMessageFormatter::formatMessage(tr("$ This client supports Auto-Completion using the Tab key.")));
+    ircBrowser->append(IrcMessageFormatter::formatMessage(tr("$ Type <b>/help</b> for commands or <b>/help [command]</b> for command syntax.")));
+    ircBrowser->append("\n");
+
+    mIrcStarted = true;
 }
 
-void dlgIRC::sendMsg()
+QPair<bool, QString> dlgIRC::sendMsg(const QString& target, const QString& message)
 {
-    QString txt = lineEdit->text();
-    lineEdit->clear();
-    if (txt.startsWith("/nick ")) {
-        txt.replace("/nick ", "");
-        connection->setNickName(txt);
-        mNick = txt;
-        connection->sendCommand(IrcCommand::createNames("#mudlet"));
-        return;
+    if (message.isEmpty()) {
+        return QPair<bool, QString>(true, "message processed by client.");
+    }
+
+    QString msgTarget = target;
+    if (target.isEmpty()) {
+        msgTarget = mChannels.first();
+    }
+
+    // inform the command parser of the target for this message.
+    // parses the message and then reverts the target to avoid confusing our UI.
+    QString lastParserTarget = commandParser->target();
+    commandParser->setTarget(msgTarget);
+    IrcCommand* command = commandParser->parse(message);
+    commandParser->setTarget(lastParserTarget);
+
+    if (!command) {
+        return QPair<bool, QString>(false, "command or message could not be parsed.");
+    }
+
+    bool sendCommand = processCustomCommand(command);
+    if (!sendCommand) {
```

Comment:
So would `bool customCommand` work in place here to indicate that we're dealing with a custom command here?